src/micro_endpoint.c

```diff
@@ -59,78 +58,86 @@ micro_new_endpoint(microEndpoint **new_ep, microService *m, const char *prefix,
     memset(&ep->stats, 0, sizeof(ep->stats));
 
+    // extra retain before subscribing since we'll need to hold it until
+    // on_complete on the subscription.
+    retain_endpoint(ep);
+
     if (ep->is_monitoring_endpoint)
-        s = natsConnection_Subscribe(&ep->sub, ep->m->nc, ep->subject, handle_request, ep);
+        s = natsConnection_Subscribe(&sub, ep->m->nc, ep->subject, handle_request, ep);
     else
-        s = natsConnection_QueueSubscribe(&ep->sub, ep->m->nc, ep->subject, MICRO_QUEUE_GROUP, handle_request, ep);
+        s = natsConnection_QueueSubscribe(&sub, ep->m->nc, ep->subject, MICRO_QUEUE_GROUP, handle_request, ep);
 
-    IFOK(s, natsSubscription_SetOnCompleteCB(ep->sub, on_drain_complete, ep));
+    if (s == NATS_OK)
+    {
+        natsSubscription_SetOnCompleteCB(sub, _release_on_complete, ep);
+
+        micro_lock_endpoint(ep);
+        ep->sub = sub;
+        ep->is_running = true;
+        micro_unlock_endpoint(ep);
+    }
+    else
+    {
+        natsSubscription_Destroy(sub); // likely always a no-op.
+        release_endpoint(ep);    // to compensate for the extra retain above.
+    }
 
     return micro_ErrorFromStatus(s);
 }
 
 microError *
-micro_stop_endpoint(microEndpoint *ep)
+micro_destroy_endpoint(microEndpoint *ep)
 {
     natsStatus s = NATS_OK;
+    natsSubscription *sub = NULL;
+    bool is_running = false;
 
-    if ((ep == NULL) || (ep->sub == NULL))
+    if (ep == NULL)
         return NULL;
 
-    if (natsConnection_IsClosed(ep->m->nc) || !natsSubscription_IsValid(ep->sub))
-    {
-        // <>/<> does the subscription need to be closed? It will be Destroy-ed
-        // when the endpoint is destroyed.
+    micro_lock_endpoint(ep);
+    sub = ep->sub;
+    is_running = ep->is_running;
+    micro_unlock_endpoint(ep);
+
+    if (!is_running)
         return NULL;
-    }
 
-    // Initiate draining the subscription. Do not release the endpoint until
-    // the on_drain_complete.
-    retain_endpoint(ep);
-    s = natsSubscription_Drain(ep->sub);
-    if (s != NATS_OK)
+    if (!natsConnection_IsClosed(ep->m->nc) && natsSubscription_IsValid(sub))
     {
-        release_endpoint(ep);
-        return microError_Wrapf(micro_ErrorFromStatus(s),
-                                "failed to stop endpoint %s: failed to drain subscription", ep->name);
+        // When the drain is complete, will release the final ref on ep.
+        s = natsSubscription_Drain(sub);
+        if (s != NATS_OK)
+        {
+            return microError_Wrapf(micro_ErrorFromStatus(s),
+                                    "failed to stop endpoint %s: failed to drain subscription", ep->name);
+        }
     }
-    return NULL;
-}
-
-microError *
-micro_destroy_endpoint(microEndpoint *ep)
-{
-    microError *err = NULL;
-    err = micro_stop_endpoint(ep);
-    if (err != NULL)
-        return err;
 
+    // Release ep since it's no longer running (to compensate for the retain in
+    // micro_start_endpoint).
```

Comment:
I don't think this is correct. If `ep->sub` is not `NULL`, then I think this will be 1 release too many. You have a single retain, but this release + the release in in `_release_on_complete`. So it would seem to me that I would do a release here only if `sub` is `NULL`, for the cases where we would call destroy in a situation where the subscription has not been created (if at all possible). But maybe I am missing something.

Reply:
@kozlovic `ep` is created with a ref count of 1, in `micro_new_endpoint` and `retain`ed in `micro_start_endpoint` so 2 releases in total are necessary to free it: one from `microService_Stop`->`micro_destroy_endpoint`, and one more from the sub's `OnComplete`. In case of the connection closing, they get executed in "reverse" oder, the subscriptions are terminated before the on_connection_closed callback, which stops the service.

Reply:
Ok, thanks.